### PR TITLE
wgengine/router/osrouter: remove orphaned tailnet addrs on cleanup

### PR DESCRIPTION
**Pull request by bcreane**
wgengine/router/osrouter: remove orphaned tailnet addrs on cleanup

The orphan-address sweep added in #20199 ran only inside Router.Set, so the teardown path (`tailscaled --cleanup`, and the unconditional cleanup at daemon start) never removed stale Tailscale addresses a previous instance left on a persistent tailscale0 - it only flushed iptables/nftables.

Wire address removal into cleanUp: with no desired config, every Tailscale-range address on the interface is an orphan, so enumerate and delete them all (IPv4 and IPv6, best-effort) in removeOrphanedAddrsForCleanup. The raw netlink/ip delete is factored out of delAddress as delAddrRaw (no v6 gate, no loopback-rule teardown, both unneeded during cleanup), and `tailscaleInterfaceAddrs` takes a keep predicate so cleanup can enumerate v6-inclusively while the Set sweep keeps its existing `reconcilableTailscaleIP` filter.

The Set-time sweep is otherwise unchanged; re-running it on network changes (netmon) for late orphans remains a follow-up (tailscale/corp#43882).

Updates #19974
Fixes tailscale/corp#44173

### CODE REVIEW

File: wgengine/router/osrouter/router_linux_test.go [hunk 1]
```diff
@@ -1933,3 +1933,57 @@ func TestSetSkipsV6OrphansWhenV6Unavailable(t *testing.T) {
 		t.Errorf("v6 orphan should be left alone when v6 is unavailable; ips=%q", fake.ips)
 	}
 }
+
+// TestCleanUpRemovesAllTailscaleAddrs verifies the teardown path
+// (removeOrphanedAddrsForCleanup, used by tailscaled --cleanup) removes every
+// Tailscale-range address -- both IPv4 CGNAT and IPv6 ULA -- while leaving
```
Comment (bradfitz) [inline wgengine/router/osrouter/router_linux_test.go]: keep in mind that people use Tailscale on machines that already use IPv4 CGNAT addresses that aren't ours. We shouldn't be bringing those ones down.

Do you have a test for that?

Comment (bcreane) [reply wgengine/router/osrouter/router_linux_test.go]: Good call-out. Three additional tests just added:

* Wrong interface - the sweep only enumerates and deletes on `dev tailscale0`. A CGNAT address that isn’t ours - e.g. ISP CGNAT on `eth0` - is never enumerated. See `TestCleanUpOnlyTouchesTunInterface`.

* On `tailscale0` but out-of-range - we use `tsaddr.IsTailscaleIP`, not a raw 100.64.0.0/10 prefix check, so the ChromeOS range, CGNAT boundaries (100.63.x/100.128.x), and ordinary link local addresses are all left alone. See `TestCleanUpKeepsNearRangeAddrs` and `TestSetKeepsNonTailscaleAddrs`.

Note I have manual scripts I’ve been running on my dev machine to verify these edge cases, but I hadn’t written unit tests, so this was a helpful comment.

File: wgengine/router/osrouter/router_linux.go [hunk 2]
```diff
@@ -1038,6 +1047,60 @@ func orphanedAddrs(kernelAddrs, desired []netip.Prefix) []netip.Prefix {
...
+}
+
+// sweepOrphanedAddrs removes Tailscale-range interface addresses that are
+// neither desired nor already tracked. It enumerates the interface, so callers
+// gate how often it runs; it does not consult or update the r.lastScanAddrs scan
+// gate. scanned reports whether the enumeration succeeded (so Set can decide
+// whether to advance its gate); errs holds any per-address deletion failures.
+// The live Set caller must hold r.mu; the teardown caller passes an unshared
+// throwaway router, so there's nothing to lock.
+func (r *linuxRouter) sweepOrphanedAddrs(s orphanSweep) (scanned bool, errs []error) {
```
Comment (sfllaw) [inline wgengine/router/osrouter/router_linux.go]: 📐 Design: I am having a lot of trouble understanding how linuxRouter.sweepOrphanedAddrs is supposed to be used. The fact that orphanSweep requires a lot of documentation when it is the argument to sweepOrphanedAddrs is not a good sign. Could this be done in a clearer (and possibly more imperative) way?

Notes:
1. Instead of `del`, could you range over an iter.Seq[netip.Prefix] and call your removal function on each element?
2. `tracked` doesn’t make sense to me, why isn’t removal idempotent, shouldn’t it always be OK to double-remove?
3. The names for `keep` and `desired` are too similar.
4. If `desired` are addresses to retain, why isn’t it a set.Set? Checking this at time-of-removal would be clearer than trying to pass it in as an argument.

Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go]: @sfllaw thank you for your review. I’m going to drop this back into _draft_ and reevaluate the approach. I agree that its not as straightforward and comprehensible as it should be.

Comment (sfllaw) [reply wgengine/router/osrouter/router_linux.go]: @bcreane Sounds good to me. Please ping me when you want me to look at it again. Or feel free to drop something on my calendar.

Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go]: @sfllaw - I tried to combine too much of the Set time change with the new Cleanup-driven code and the result was an unreadable mess. This version has a little more redundant code, and should be much more readable (it leaves the Set changes basically alone). 

With respect to your notes:
- "can't understand sweepOrphanedAddrs/orphanSweep" -> both deleted; one caller, no struct.
- "range over a seq and call remove per element instead of del" -> that's exactly the loop now.
- "tracked doesn't make sense" -> gone from cleanup entirely (only main's pre-existing prevAddrs skip remains in the Set path).
- "keep/desired names too similar" -> no desired anymore.
- "why not a set.Set checked at removal" -> cleanup has no retain-set because we remove every Tailscale-range address on the tunnel interface.

File: wgengine/router/osrouter/router_linux.go [hunk 3]
```diff
@@ -936,22 +936,34 @@ func (r *linuxRouter) delAddress(addr netip.Prefix) error {
...
+	if err := netlink.AddrDel(link, nlAddrOfPrefix(addr)); err != nil {
+		return fmt.Errorf("deleting address %v from tunnel interface: %w", addr, err)
 	}
 	return nil
 }
 
+// tailscaleRangeIP reports whether ip is within Tailscale's address ranges.
+func tailscaleRangeIP(ip netip.Addr) bool {
+	return tsaddr.IsTailscaleIP(ip.Unmap())
+}
```
Comment (sfllaw) [inline wgengine/router/osrouter/router_linux.go]: 🐞 Bug (blocking): I think this actually a bug in [tsaddr.IsTailscaleIP](https://pkg.go.dev/tailscale.com/net/tsaddr#IsTailscaleIP)? I haven’t looked deeply at corp, but under cursory examination, none of the callers could break if we always unmapped. And the implementation of [netip.Addr.Unmap](https://pkg.go.dev/net/netip#Addr.Unmap) is really cheap.

``` go
// IsTailscaleIP reports whether IP is an IP address in a range that
// Tailscale assigns from.
func IsTailscaleIP(ip netip.Addr) bool {
	ip = ip.Unmap()
	return IsTailscaleIPv4(ip) || TailscaleULARange().Contains(ip)
}

// IsTailscaleIPv4 reports whether an IPv4 IP is an IP address that
// Tailscale assigns from.
// It will always return false if ip is an "IPv4-mapped IPv6 address".
func IsTailscaleIPv4(ip netip.Addr) bool {
	return CGNATRange().Contains(ip) && !ChromeOSVMRange().Contains(ip)
}
```

Note that corp _does_ rely on [IsTailscaleIPv4 being IPv4-only](https://github.com/search?q=repo%3Atailscale%2Fcorp+%22tsaddr.IsTailscaleIPv4%22+language%3AGo&type=code), which is why I added that extra bit of documentation.

File: wgengine/router/osrouter/router_linux.go
Comment (sfllaw) [inline wgengine/router/osrouter/router_linux.go]: 📛 Naming (non-blocking): After your refactoring, could this be renamed to `IsDeletableAddr`? I think that’s its actual intent.

File: wgengine/router/osrouter/router_linux.go [hunk 4]
```diff
@@ -964,13 +976,11 @@ func (r *linuxRouter) reconcilableTailscaleIP(ip netip.Addr) bool {
...
 }
 
-// tailscaleInterfaceAddrs returns the addresses on the tunnel interface that are
-// within Tailscale's ranges and that the router can act on (see
-// reconcilableTailscaleIP); all others are excluded so they're never candidates
-// for removal. It's a filtered subset, and errors if the interface can't be read.
-func (r *linuxRouter) tailscaleInterfaceAddrs() ([]netip.Prefix, error) {
+// tailscaleInterfaceAddrs returns the addresses on the tunnel interface for
+// which keep reports true. It errors if the interface can't be read.
+func (r *linuxRouter) tailscaleInterfaceAddrs(keep func(netip.Addr) bool) ([]netip.Prefix, error) {
```
Comment (sfllaw) [inline wgengine/router/osrouter/router_linux.go]: 📐 Design (blocking): This method conflates two operations: enumerating all the addresses on the tailscale tun interface and then filtering for the IP addresses on the tailnet. You’ve tried to make this more generic by passing in a filter function, but I think it would be cleaner if we broke it out. That’s why I suggested that you return [iter.Seq](https://pkg.go.dev/iter#Seq):

```suggestion
// tailscaleInterfaceAddrs returns the addresses on the Tailscale tunnel interface.
// It errors if the interface can't be read.
func (r *linuxRouter) tailscaleInterfaceAddrs() (iter.Seq[netip.Prefix], error) {
```

Instead of the `keep` closures, you can use these iterators in conjuction with [slices.Collect](https://pkg.go.dev/slices#Collect):
```go
// tailscaleAddrs yields only Tailscale addresses,
// according to [tsaddr.IsTailscaleIP].
func tailscaleAddrs(addrs iter.Seq[netip.Prefix]) iter.Seq[netip.Prefix] {
	return func(yield func(netip.Prefix) bool) {
		for p := range addrs {
			if tsaddr.IsTailscaleIP(p.Addr()) && !yield(p) {
				return
			}
		}
	}
}

// deletableAddrs yields addresses that could be deleted,
// according to [linuxRouter.isDeletableAddr].
func (r *linuxRouter) deletableAddrs(addrs iter.Seq[netip.Prefix]) iter.Seq[netip.Prefix] {
	return func(yield func(netip.Prefix) bool) {
		for p := range addrs {
			if r.isDeletableAddr(p.Addr()) && !yield(p) {
				return
			}
		}
	}
}
```

File: wgengine/router/osrouter/router_linux.go [hunk 5]
```diff
@@ -936,22 +936,34 @@ func (r *linuxRouter) delAddress(addr netip.Prefix) error {
 	if err := r.delLoopbackRule(addr.Addr()); err != nil {
 		return err
 	}
+	return r.delAddrRaw(addr)
+}
+
+// delAddrRaw deletes addr from the tunnel interface without the v6-availability
+// gate or loopback-rule teardown that delAddress does. delAddress wraps it;
+// removeOrphanedAddrsForCleanup uses it directly (see there).
+func (r *linuxRouter) delAddrRaw(addr netip.Prefix) error {
```
Comment (sfllaw) [inline wgengine/router/osrouter/router_linux.go:1033]: 🐞 Bug (blocking): It seems like there’s a bug here, where a missing firewall rule can leak the address. Shouldn’t we make delAddress idempotent and have it try to remove the address no matter what? Since the callers only ever log errors, I would suggest removing the IPv6 check so that we log whenever we’ve added a v6 address but can’t delete it.

```go
// delAddress removes an IP/mask from the tunnel interface. Fails if
// the address is not assigned to the interface, or if the removal
// fails.
func (r *linuxRouter) delAddress(addr netip.Prefix) error {
	var errs []error
	if err := r.delLoopbackRule(addr.Addr()); err != nil {
		errs = append(errs, err)
	}
	if err := r.delAddrRaw(addr); err != nil {
		errs = append(errs, err)
	}
	return errors.Join(errs...)
}
```

If I were extra keen, I might suppress any error messages about missing addrs inside delLoopbackRule or delAddrRaw. But this should only happen if something raced us to the deletion, which should be quite exceptional.

File: wgengine/router/osrouter/router_linux.go [hunk 6]
```diff
@@ -1824,10 +1834,48 @@ func platformCanNetfilter() bool {
...
+	// tailscaleRangeIP + delAddrRaw remove IPv6 orphans even when this process
+	// never brought IPv6 up, unlike the v6-gated live sweep.
+	addrs, err := r.tailscaleInterfaceAddrs(tailscaleRangeIP)
+	if err != nil {
+		r.logf("enumerating %s addresses for cleanup failed: %v", interfaceName, err)
+		return
+	}
+	removed := make([]netip.Prefix, 0, len(addrs))
+	for _, p := range addrs {
+		if err := r.delAddrRaw(p); err != nil {
```
Comment (sfllaw) [inline wgengine/router/osrouter/router_linux.go]: ⬆️ Use an idempotent delAddress instead, see https://github.com/tailscale/tailscale/pull/20304#discussion_r3597868945:

```suggestion
		if err := r.delAddress(p); err != nil {
```

**Review by sfllaw — CHANGES_REQUESTED** [review]
Thanks for doing the refactoring.

Now that I understand the logic better, I think I’ve identified a couple of bugs? Let me know what you think.

File: wgengine/router/osrouter/router_linux.go (hunk 2 above)
Comment (sfllaw) [reply wgengine/router/osrouter/router_linux.go]: Thanks for the massive cleanup. Sorry it took me a while to review this, but it took me a while to read and understand. Here’s my new review: https://github.com/tailscale/tailscale/pull/20304#pullrequestreview-4715894660.

File: wgengine/router/osrouter/router_linux.go (hunk 3 above)
Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go]: Good catch, yup it's a bug in `tsaddr.IsTailscaleIP`: the ip.Is4() branch sends IPv4-mapped IPv6 addresses (e.g. `::ffff:100.64.0.1`) down the ULA-only path, so a mapped CGNAT address wrongly returns false. As you suggested, will fix by unmapping at the top of `IsTailscaleIP`. Adding some UTs as well.

File: wgengine/router/osrouter/router_linux.go
Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go]: will do

File: wgengine/router/osrouter/router_linux.go (hunk 4 above)
Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go]: Separating concerns is a valuable suggestion - the code is easier to reason about now that enumeration and filtering are distinct. I wasn't familiar with the `iter.Seq/slices.Collect` pattern; I'm not sure it buys much here (it boils down to a slice anyway), but it's the pattern laid down elsewhere in the repo, so I'll take it.

Pending change: split into `tailscaleInterfaceAddrs` (enumerate) plus `tailscaleAddrs/deletableAddrs` filters.

File: wgengine/router/osrouter/router_linux.go (hunk 5 above)
Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go:1033]: Another good catch - and a slightly painful one, because it's the same shape as the bug I just fixed in [#20488](https://github.com/tailscale/tailscale/pull/20488/changes#diff-333c95b8c44f059ab890627b0f8f8cf323d82f42e86e03a561a51ece0c4ef836L919-L942): an early return on one failure skipping a later required action. In the earlier case it was `router.Set` failing and skipping DNS configuration; and here a `delLoopbackRule` error returning early and skipping the address delete, leaking the address. Bummer to reintroduce the exact pattern I'd just cleaned up - I should have spotted it.

Fixed as you suggested: `delAddress` now attempts both the loopback-rule teardown and the address delete regardless, joins their errors with `errors.Join`, and no longer no-ops on IPv6 - callers only log, and the cleanup path relies on v6 orphans actually being deleted. The cleanup sweep now goes through the idempotent delAddress (last comment) rather than the raw delete.

File: wgengine/router/osrouter/router_linux.go (hunk 6 above)
Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go]: will do

**Comment by bcreane** [comment]
@sfllaw thank you for your detailed and thoughtful review! I thought I was starting to get the hang of this area if the code base, but feels like there's a ways to go still. 

I need to do some more manual tests, but UTs are all passing, and I think I implemented all of your fixes / suggestions with fidelity.

File: wgengine/router/osrouter/router_linux.go (hunk 4 above)
Comment (sfllaw) [reply wgengine/router/osrouter/router_linux.go]: In theory, you could also have orphanedAddrs return an iter.Seq and then you would probably save memory by not allocating throw-away slices. Since the iterators are so simple, those tests will get inlined by the compiler in the `for ... range orphaned` loop. Sure, you won’t be able to preallocate the `removed` slice, but I would wager that it is small (or even nil) most of the time.

```go
if ifaceAddrs, err := r.tailscaleInterfaceAddrs(); err != nil {
	r.logf("router: enumerating interface addresses failed, skipping orphan cleanup: %v", err)
} else {
	r.lastScanAddrs = wantAddrs
	kernelAddrs := r.deletableAddrs(ifaceAddrs)
	var removed []netip.Prefix
	for p := range orphanedAddrs(kernelAddrs, cfg.LocalAddrs) {
		if prevAddrs[p] {
			continue // an address we were tracking; cidrDiff already handled it
		}
		if err := r.delAddress(p); err != nil {
			r.logf("router: removing stale address %v from %s failed: %v", p, r.tunname, err)
			errs = append(errs, err)
			continue
		}
		removed = append(removed, p)
	}
	if len(removed) > 0 {
		r.logf("router: removed %d stale Tailscale address(es) from %s left by a previous instance: %v", len(removed), r.tunname, removed)
	}
}
```

**Review by sfllaw — APPROVED** [review]
I have no more blocking concerns, so approved.

You may want to consider converting the ophanedAddrs method to return an iter.Seq as well: https://github.com/tailscale/tailscale/pull/20304#discussion_r3599830697

File: wgengine/router/osrouter/router_linux.go (hunk 4 above)
Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go]: smart suggestion, will do.

File: wgengine/router/osrouter/router_linux.go [hunk 7]
```diff
@@ -1012,49 +1014,80 @@ func (r *linuxRouter) addAddress(addr netip.Prefix) error {
...
 
-// reconcilableTailscaleIP reports whether ip, found on the tunnel interface, is
-// a Tailscale-range address the router can actually delete. delAddress no-ops on
-// IPv6 when IPv6 is unavailable, so excluding those avoids treating a no-op
-// "deletion" as success.
-func (r *linuxRouter) reconcilableTailscaleIP(ip netip.Addr) bool {
+// isDeletableAddr reports whether ip, found on the tunnel interface, is a
+// Tailscale-range address the live router should delete. It excludes IPv6 when
+// IPv6 is unavailable so the Set-time sweep doesn't churn on addresses this
+// instance couldn't have installed; the teardown path uses tailscaleAddrs
```
Comment (sfllaw) [inline wgengine/router/osrouter/router_linux.go]: 📄 Documentation (non-blocking): Consider adding a doc link so that your code editor can jump to the definition:

```suggestion
// instance couldn't have installed; the teardown path uses [tailscaleAddrs]
```

File: wgengine/router/osrouter/router_linux.go [hunk 8]
```diff
@@ -1104,24 +1132,24 @@ func (r *linuxRouter) tailscaleInterfaceAddrsIPCommand() ([]netip.Prefix, error)
...
-			if r.reconcilableTailscaleIP(p.Addr()) {
-				ret = append(ret, p)
-			}
+			ret = append(ret, p)
 			break
 		}
 	}
-	return ret, nil
+	return slices.Values(ret), nil
 }
```
Comment (sfllaw) [inline wgengine/router/osrouter/router_linux.go:1140]: 📐 Design (non-blocking): Consider refactoring this method so that the iterator pulls results instead of collecting them in the `ret` slice:

```go
// tailscaleInterfaceAddrsIPCommand is the "ip" command implementation of
// tailscaleInterfaceAddrs, used in tests and when TS_DEBUG_USE_IP_COMMAND is
// set.
func (r *linuxRouter) tailscaleInterfaceAddrsIPCommand() (iter.Seq[netip.Prefix], error) {
	out, err := r.cmd.output("ip", "-oneline", "addr", "show", "dev", r.tunname)
	if err != nil {
		return nil, err
	}
	return func(yield func(netip.Prefix) bool) {
		for line := range bytes.Lines(out) {
			// `ip -oneline addr show` puts each address on one line as
			// "inet <cidr>" or "inet6 <cidr>".
			fields := strings.Fields(string(line))
			for i := 0; i+1 < len(fields); i++ {
				if fields[i] != "inet" && fields[i] != "inet6" {
					continue
				}
				p, err := netip.ParsePrefix(fields[i+1])
				if err != nil {
					break
				}
				if !yield(p) {
					return
				}
				break
			}
		}
	}, nil
}
```

This should also work for tailscaleInterfaceAddrs above.

File: net/tsaddr/tsaddr.go [hunk 9]
```diff
@@ -78,6 +79,7 @@ func IsTailscaleIP(ip netip.Addr) bool {
 
 // IsTailscaleIPv4 reports whether an IPv4 IP is an IP address that
 // Tailscale assigns from.
+// It will always return false if ip is an "IPv4-mapped IPv6 address".
```
Comment (sfllaw) [inline net/tsaddr/tsaddr.go:82]: 🧪 Testing (non-blocking): Is there a test case for this? If not, might I suggest adding one to TestIsTailscaleIPv4 for `netip.AddrFrom16(netip.MustParseAddr("100.67.19.57").As16())` to ensure we don’t break it.

File: wgengine/router/osrouter/router_linux.go (hunk 7 above)
Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go]: will do

File: net/tsaddr/tsaddr.go (hunk 9 above)
Comment (bcreane) [reply net/tsaddr/tsaddr.go:82]: will do

File: wgengine/router/osrouter/router_linux.go (hunk 8 above)
Comment (bcreane) [reply wgengine/router/osrouter/router_linux.go:1140]: I'll keep this as a future improvement. Thanks for the suggestion.

**Review by sfllaw — APPROVED** [review]
Approved without reservations. Thanks for the cleanup and bugfix.
